includes/class-amp-story-post-type.php

```diff
@@ -341,7 +343,271 @@ public static function remove_classic_editor_link( $actions, $post ) {
+					$found    = true;
+					$url_attr = in_array( $css_selector, $css_url_data_types, true );
+				}
+			}
+
+			if ( $found && $url_attr ) {
+				// Simplified: matches the sequence `url(*)`.
+				preg_match_all( '/url\([^)]+\)/', $parts[1], $url_matches );
+
+				foreach ( $url_matches[0] as $url_match ) {
+					// Clean up the URL from each of the matches above.
+					preg_match( '/^url\(\s*([\'\"]?)(.*)(\g1)\s*\)$/', $url_match, $url_pieces );
+
+					if ( empty( $url_pieces[2] ) ) {
+						$found = false;
+						break;
+					}
+
+					$url = trim( $url_pieces[2] );
+
+					if ( empty( $url ) || wp_kses_bad_protocol( $url, $allowed_protocols ) !== $url ) {
+						$found = false;
+						break;
+					} else {
+						// Remove the whole `url(*)` bit that was matched above from the CSS.
+						$css_test_string = str_replace( $url_match, '', $css_test_string );
+					}
+				}
+			}
+
+			if ( $found ) {
+				if ( '' !== $css ) {
+					$css .= ';';
+				}
+
+				$css .= $css_item;
+			}
+		}
+
+		return $css;
+	}
+
+	/**
+	 * Filters the response before executing any REST API callbacks.
+	 *
+	 * Temporarily modifies post content during saving in a way that KSES
+	 * does not strip actually valid CSS from post content, making block content invalid.
+	 *
+	 * @todo Remove once core has better CSS parsing.
+	 *
+	 * @link https://core.trac.wordpress.org/ticket/37134
+	 *
+	 * @param WP_HTTP_Response|WP_Error $response Result to send to the client. Usually a WP_REST_Response or WP_Error.
+	 * @param array                     $handler  Route handler used for the request.
+	 * @param WP_REST_Request           $request  Request used to generate the response.
+	 *
+	 * @return WP_HTTP_Response|WP_Error The filtered response.
+	 */
+	public static function filter_rest_request_for_kses( $response, $handler, $request ) {
+		if ( is_wp_error( $response ) ) {
+			return $response;
+		}
+
+		$search = '\/' . self::POST_TYPE_SLUG . '\/';
+
+		$editable_request_methods = array_map( 'trim', explode( ',', WP_REST_Server::EDITABLE ) );
+
+		if ( ! in_array( $request->get_method(), $editable_request_methods, true ) || ! preg_match( "/{$search}/i", $request->get_route() ) ) {
+			return $response;
+		}
+
+		if ( ! current_user_can( 'edit_post', $request['id'] ) ) {
+			return $response;
+		}
+
+		$style_attr_values = [];
+
+		// Replace inline styles before KSES...
+		add_filter(
+			'content_save_pre',
+			static function ( $post_content ) use ( &$style_attr_values ) {
+				$post_content = preg_replace_callback(
+					'|<\w+(?:-\w+)*\s[^>]*?style=\\\"([^"]*)\\\"[^>]+?>|', // Because the post data is slashed.
+					static function ( $matches ) use ( &$style_attr_values ) {
+						$hash                       = md5( $matches[1] );
+						$style_attr_values[ $hash ] = self::safecss_filter_attr( $matches[1] );
+
+						// Replaces the complete style attribute value with its hashed version.
+						return str_replace( $matches[1], $hash, $matches[0] );
```

Comment:
To be a little more robust, I think the pattern should be tweaked a bit more as:

```php
'|(?P<before><\w+(?:-\w+)*\s[^>]*?)style=\\\"(?P<styles>[^"]*)\\\"(?P<after>[^>]+?>)|'
```

Then instead of doing a replacement in the string of `$matches` it can just concatenate:

```php
return $matches['before'] . sprintf( ' style="%s" ', $hash ) . $matches['after'];
```

What's more: to guard against a future `safecss_filter_attr()` which could sanitize away the md5 hash, what if the hash gets placed inside of a Kses-allowed data attribute like `data-temp-style-hash`. Then this could be:

```php
return $matches['before'] . sprintf( ' data-temp-style-hash="%s" ', $hash ) . $matches['after'];
```

And the replacement below could be:

```php
return preg_replace_callback(
    '/ data-temp-style-hash="(?P<hash>[0-9a-f]+)" /',
    function ( $matches ) use ( $style_attr_values ) {
        return isset( $style_attr_values[ $matches['hash'] ] ) ? sprintf( ' style="%s" ', esc_attr( wp_slash( $style_attr_values[ $matches['hash'] ] ) ) ) : '';
    },
    $post_content
);
return str_replace( array_keys( $style_attr_values ), array_values( $style_attr_values ), $post_content );
```